### PR TITLE
Update debian scripts for debhelper-compat 12

### PR DESCRIPTION
**Pull request by callahad**
This is part removing support for Debian Stretch and Ubuntu Xenial
following the end of life for Python 3.5.

Other notable changes:

- Architecture is set to `any` to allow (but not mandate) ARM builds

- The dh-virtualenv Dockerfile was overhauled to more closely follow the
  example provided upstream, including a necessary tweak to ensure
  builds work on Ubuntu bionic.

Signed-off-by: Dan Callahan <danc@element.io>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

**Comment by callahad** [comment]
Looking at the artifacts produced for Buster (and excluding expected changes like timestamps and checksums, eggs vs wheels), here's what's different:

- `control/prerm`, `control/postrm`, and `control/postinst` only call `deb-systemd-invoke` or `systemctl`, they do not _also_ call `invoke-rc.d` or `update-rc.d`
- `control/conffiles` no longer lists `/etc/default/matrix-synapse` and it is not present in the data files. Its content was:

  ```sh
  # Specify environment variables used when running Synapse
  # SYNAPSE_CACHE_FACTOR=0.5 (default)
  ```

Otherwise, everything else is effectively identical. All other modifications are as seen in this pull request.

File: debian/rules
Comment (anoadragon453) [inline debian/rules:54]: Note that including comments here, rather than above the `%:`, will cause them to be printed out when the deb is built, e.g.:

```
dpkg-buildpackage: info: host architecture amd64
 dpkg-source --before-build build
 debian/rules clean
# TODO: Experiment with --buildsystem=dh_virtualenv
# https://dh-virtualenv.readthedocs.io/en/latest/usage.html#experimental-buildsystem-support
dh clean --with python-virtualenv --buildsystem=pybuild
   dh_clean -O--buildsystem=pybuild
 dpkg-source -b build
```

**Review by anoadragon453 — COMMENTED** [review]
I'll admit that I'm not very confident reviewing deb build scripts, but nothing added here looks inherently wrong.

Attempting to build the image and then a `.deb` with that image using `distro=bionic` worked flawlessly on my machine.

```sh
$ docker build --build-arg distro=ubuntu:bionic -t callahad/synapse-deb -f docker/Dockerfile-dhvirtualenv .
$ docker run -v /path/to/synapse/checkout:/synapse/source -v /output/deb/dir:/debs callahad/synapse-deb
```

File: docker/Dockerfile-dhvirtualenv
Comment (callahad) [reply docker/Dockerfile-dhvirtualenv:79]: No clue, but minimizes diff [against the upstream Dockerfile](https://github.com/spotify/dh-virtualenv/commit/af7669b5dfc92c669b52afd8be53d17b1e99ef6e)

**Comment by callahad** [comment]
SyTest flaked on, re-running. Otherwise everything is green.

```
 not ok 463 User can invite remote user to room with version 5
# Started: 2021-04-12 14:34:34.416
# Ended: 2021-04-12 14:34:44.563
# Timed out waiting for test at /sytest/run-tests.pl line 771.
# 0.113365: Registered new user @anon-20210412_142620-550:localhost:8875
# 0.143863: Registered new user @anon-20210412_142620-549:localhost:8837
# 1.070622: Invited user @anon-20210412_142620-550:localhost:8875 to !aGuqpsxlGstWktTddm:localhost:8837
# {}
```

**Comment by callahad** [comment]
> `control/conffiles` no longer lists `/etc/default/matrix-synapse` and it is not present in the data files.

Looks like this is to be expected with the move to just using `dh_installsystemd`. I've modified the service files to ensure that they don't fail if the file is absent, but do use it if present. Which should make this painless for new and existing installs alike.

**Comment by callahad** [comment]
Another flakey sytest, on the `(buster, multi-postgres, workers)` job

```
 not ok 361 A prev_batch token from incremental sync can be used in the v1 messages API
# Started: 2021-04-12 20:35:24.566
# Ended: 2021-04-12 20:35:25.432
# Expected only one timeline event at tests/31sync/04timeline.pl line 409.
# 0.093688: Registered new user @anon-20210412_203031-355:localhost:8837
# 0.585465: Sent 'm.room.message' event in !rpgWuoQAgpoHSdirvC:localhost:8837: $6nFdKD6RhcRrpdvDo_bNI-vc3UOEa16Vcma0357-50Q
# 0.811684: Sent 'm.room.message' event in !rpgWuoQAgpoHSdirvC:localhost:8837: $n_k0FUKxwNOA17JoTIS8GJNaxF11CwPem4K10UMY9H0
```

**Comment by callahad** [comment]
This time the flakey SyTest was on `buster, postgres, workers, redis` with

```
not ok 411 Old leaves are present in gapped incremental syncs
# Started: 2021-04-12 22:01:27.515
# Ended: 2021-04-12 22:01:30.054
# Expected only 2 membership events at tests/10apidoc/09synced.pl line 396.
# 0.193599: Registered new user @anon-20210412_215424-436:localhost:8837
# 0.194552: Registered new user @anon-20210412_215424-438:localhost:8837
# 0.197405: Registered new user @anon-20210412_215424-437:localhost:8837
# 1.131526: User @anon-20210412_215424-437:localhost:8837 joined room
# { room_id => "!slNjFXEVPPvIjYdNGZ:localhost:8837" }
# 1.425742: User @anon-20210412_215424-438:localhost:8837 joined room
# { room_id => "!slNjFXEVPPvIjYdNGZ:localhost:8837" }
# 1.582264: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $_nxmSX5Wf1KXxQWjoeA8sOusPgXLuWGoRQu59qB8pdY
# 1.707723: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $IqLg65UIhtuviyXeWr_9bbw9rVNpr4xmRFpNZ4NP2Y0
# 1.816230: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $0j3azy3SoIrw1s5ZsdAAaSMWKWG3bhS30nmctpfmGHM
# 1.894438: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $Q-Cc_FeRnOHB5c-W20nieDoOaBPsbenqItRdU_NNUv4
# 1.971960: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $jMCvUkDufR4AzsHKc7feS1gOteY-DKYycolivRNgaaA
# 2.047663: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $g-BOm2Bxc3TXvO6rKP8cNfuC-FvmJbj_cYrZWC5OOZk
# 2.145319: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $GPfV5UHithCgKY87Mrn3PPcNmPhj_lhZl5reSOTzbSE
# 2.198933: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $eD4GzS7z8s8ORkd8-79QWZwWXJNb6Tl00JgkiSYEJ1o
# 2.271701: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $iTuqHBQTMLNsmnXjaqYHdq8ZdNJpEMBBI3iCiflof-g
# 2.348262: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $buhyvWzM4nI3FVETEvMv_qh4kRwgVYVKHNRO-BYElAI
# 2.415376: Sent 'm.room.message' event in !slNjFXEVPPvIjYdNGZ:localhost:8837: $-8j6SFXCkQnUAYH8j0i58d3Oxd_1eN7UDbMxYtOLd04
# 2.536207: assert_state_room_members_match: expected members:
# [
#   "\@anon-20210412_215424-436:localhost:8837",
#   "\@anon-20210412_215424-437:localhost:8837",
# ]
# 2.536374: assert_state_room_members_match: actual state:
# [
#   {
#     content          => { join_rule => "public" },
#     event_id         => "\$jfulufyojdqrXRHh-KDn2E4G-ijXcq3pal1dJUGKp-Q",
#     origin_server_ts => JSON::number(1618264888049),
#     sender           => "\@anon-20210412_215424-436:localhost:8837",
#     state_key        => "",
#     type             => "m.room.join_rules",
#     unsigned         => { age => JSON::number(1992) },
#   },
#   {
#     content          => {
#                           creator => "\@anon-20210412_215424-436:localhost:8837",
#                           room_version => 6,
#                         },
#     event_id         => "\$Bq9z_YwWxxZ1yEPgwLBzoDD7Lf_tXd7an71OuzgeMnw",
#     origin_server_ts => JSON::number(1618264887766),
#     sender           => "\@anon-20210412_215424-436:localhost:8837",
#     state_key        => "",
#     type             => "m.room.create",
#     unsigned         => { age => JSON::number(2275) },
#   },
#   {
#     content          => { name => "A room name" },
#     event_id         => "\$MV687TG7hNPp9DvUd35XqaRdvvUVbF-fyqkyupW6hT4",
#     origin_server_ts => JSON::number(1618264888435),
#     sender           => "\@anon-20210412_215424-436:localhost:8837",
#     state_key        => "",
#     type             => "m.room.name",
#     unsigned         => { age => JSON::number(1606) },
#   },
#   {
#     content          => { displayname => "anon-20210412_215424-436", membership => "join" },
#     event_id         => "\$T5x_3jppujKLFpm0-u4uZwbiH3VBUavsYg8S03pgNlM",
#     origin_server_ts => JSON::number(1618264887852),
#     sender           => "\@anon-20210412_215424-436:localhost:8837",
#     state_key        => "\@anon-20210412_215424-436:localhost:8837",
#     type             => "m.room.member",
#     unsigned         => { age => JSON::number(2189) },
#   },
#   {
#     content          => { displayname => "anon-20210412_215424-437", membership => "join" },
#     event_id         => "\$c78lgnGsL_DTTWE_BveuZf2S93TBT2_iDt2IDcv0kEY",
#     origin_server_ts => JSON::number(1618264888595),
#     sender           => "\@anon-20210412_215424-437:localhost:8837",
#     state_key        => "\@anon-20210412_215424-437:localhost:8837",
#     type             => "m.room.member",
#     unsigned         => { age => JSON::number(1446) },
#   },
#   {
#     content          => {
#                           ban => JSON::number(50),
#                           events => {
#                             "m.room.avatar"             => JSON::number(50),
#                             "m.room.canonical_alias"    => JSON::number(50),
#                             "m.room.encryption"         => JSON::number(100),
#                             "m.room.history_visibility" => JSON::number(100),
#                             "m.room.name"               => JSON::number(50),
#                             "m.room.power_levels"       => JSON::number(100),
#                             "m.room.server_acl"         => JSON::number(100),
#                             "m.room.tombstone"          => JSON::number(100),
#                           },
#                           events_default => JSON::number(0),
#                           invite => JSON::number(50),
#                           kick => JSON::number(50),
#                           redact => JSON::number(50),
#                           state_default => JSON::number(50),
#                           users => {
#                             "\@anon-20210412_215424-436:localhost:8837" => JSON::number(100),
#                           },
#                           users_default => JSON::number(0),
#                         },
#     event_id         => "\$2fDkgyJFTOTWQkhD0-2h36ltD3XE_DchGSdAm6_YHLo",
#     origin_server_ts => JSON::number(1618264887925),
#     sender           => "\@anon-20210412_215424-436:localhost:8837",
#     state_key        => "",
#     type             => "m.room.power_levels",
#     unsigned         => { age => JSON::number(2116) },
#   },
#   {
#     content          => { displayname => "anon-20210412_215424-438", membership => "join" },
#     event_id         => "\$ZATUuFA2VpUTNJHlTXqWf533EAqiLDdTfhEAd8LJQ50",
#     origin_server_ts => JSON::number(1618264888880),
#     sender           => "\@anon-20210412_215424-438:localhost:8837",
#     state_key        => "\@anon-20210412_215424-438:localhost:8837",
#     type             => "m.room.member",
#     unsigned         => { age => JSON::number(1161) },
#   },
#   {
#     content          => { history_visibility => "shared" },
#     event_id         => "\$XlnVNnFd6nB_AX4aGXcxBir0zGff2W7sFXcfFW6eIhs",
#     origin_server_ts => JSON::number(1618264888156),
#     sender           => "\@anon-20210412_215424-436:localhost:8837",
#     state_key        => "",
#     type             => "m.room.history_visibility",
#     unsigned         => { age => JSON::number(1885) },
#   },
# ]
```

File: debian/rules
Comment (richvdh) [inline debian/rules:55]: note to my future self: previously we used the `python_distutils` build system; switching to `pybuild` matches the recommendation at https://wiki.debian.org/Python/LibraryStyleGuide. I don't think there was a good reason why we used `python_distutils`, other than that it was the default.

File: docker/Dockerfile-dhvirtualenv
Comment (richvdh) [inline docker/Dockerfile-dhvirtualenv:24]: I'm giving `--force-unsafe-io` some side-eye. The warnings on `dpkg` sound a bit dire: are you sure it is safe?

File: docker/Dockerfile-dhvirtualenv
Comment (richvdh) [inline docker/Dockerfile-dhvirtualenv:27]: are we basically hardcoding the build-deps of dh-virtualenv here? I'm not convinced that is an improvement, since it means we now need to keep it in sync with dh-virtualenv.

File: docker/Dockerfile-dhvirtualenv
Comment (richvdh) [inline docker/Dockerfile-dhvirtualenv:56]: this was after most of the apt-gets to help with faster iteration, but :man_shrugging:

File: docker/Dockerfile-dhvirtualenv
Comment (richvdh) [reply docker/Dockerfile-dhvirtualenv:24]: I see that dh-virtualenv's Dockerfile does it, and that we already do it below, so :man_shrugging: I guess - though I'd like to know if it is written down anywhere that this is an ok thing to do when building images.

File: debian/matrix-synapse.service
Comment (callahad) [reply debian/matrix-synapse.service:8]: We still suggest people use it in our docs, so we might as well still honor it if present. Plus, existing installations may depend on its behavior.

Systemd upstream considers `EnvironmentFile` as hack that would ideally go away, but there's no clear replacement other than directly editing the Synapse config or overriding the systemd unit file locally.

File: docker/Dockerfile-dhvirtualenv
Comment (callahad) [reply docker/Dockerfile-dhvirtualenv:24]: It's totally fine. All it does is disable fsync calls, which can be a problem if a real system with a persistent disk crashes midway through installation, but in a Docker container we either succeed and snapshot filesystem in its entirety a new layer, or we fail and everything goes away.

**Comment by richvdh** [comment]
> Looking at the artifacts produced for Buster (and excluding expected changes like timestamps and checksums, eggs vs wheels), here's what's different:
> 
>     * `control/prerm`, `control/postrm`, and `control/postinst` only call `deb-systemd-invoke` or `systemctl`, they do not _also_ call `invoke-rc.d` or `update-rc.d`

Also: `preinst` is removed. per my review: I'm not sure that is safe.

File: docker/Dockerfile-dhvirtualenv
Comment (callahad) [reply docker/Dockerfile-dhvirtualenv:27]: Yep, but 1) upstream updates are rather rare and 2) it's far easier to track down discrepancies this way (the `buster` builds failing took me a while to sort out as a result).

It also avoids another call out to the network for `mk-build-deps` and lets us get all of that out of the way (and cached) up front.

File: docker/Dockerfile-dhvirtualenv
Comment (callahad) [reply docker/Dockerfile-dhvirtualenv:56]: Sure, but moving it up here lets us remove an entire call to `apt-get update && apt-get install`, so I'd expect this to still be faster than the previous setup. And this is just dh-virtualenv itself, which shouldn't need to change much (ever?) between builds.

File: debian/matrix-synapse.service
Comment (richvdh) [reply debian/matrix-synapse.service:8]: mostly I feel like if we're going to leave the `EnvironmentFile` directive in place, we should probably also leave an example file there too. In my experience people get scared when they discover a config file doesn't exist at all.

I don't feel too strongly about it though.

File: debian/matrix-synapse-py3.triggers
Comment (richvdh) [reply debian/matrix-synapse-py3.triggers:7]: (or build this dynamically?)

**Comment by callahad** [comment]
> Also: `preinst` is removed. per my review: I'm not sure that is safe.

Yep, but that's part of the pull request itself -- that comment was trying to call attention to the _additional_ changes in the generated artifacts above and beyond what's evident in the PR :)

File: debian/matrix-synapse-py3.preinst
Comment (callahad) [reply debian/matrix-synapse-py3.preinst:31]: The comment indicates that this is a problem with the py2 synapse package, which is over two years old, so I'd not expect to see upgrades from that any longer.

Give me a minute to re-read how the maintainer scripts work...

Comment (callahad) [reply debian/matrix-synapse-py3.preinst:31]: Looking at the flowcharts in [the Debian policy manual](https://www.debian.org/doc/debian-policy/ap-flowcharts.html), a successful upgrade would invoke:

1. old `prerm upgrade`
2. new `preinst upgrade`
3. old `postrm upgrade`
4. new `postinst configure`

The `prerm` script calls `deb-systemd-invoke stop 'matrix-synapse.service' >/dev/null || true`, so doing the same thing in `preinst` appears to be redundant.

File: docker/Dockerfile-dhvirtualenv
Comment (richvdh) [reply docker/Dockerfile-dhvirtualenv:24]: fair enough!

File: debian/matrix-synapse-py3.preinst
Comment (richvdh) [reply debian/matrix-synapse-py3.preinst:31]: > The prerm script calls deb-systemd-invoke stop 'matrix-synapse.service' >/dev/null || true, so doing the same thing in preinst appears to be redundant.

the problem is/was that the prerm in matrix-synapse *didn't* do that (for the reasons set out at lines 16-20).

Comment (richvdh) [reply debian/matrix-synapse-py3.preinst:31]: > The comment indicates that this is a problem with the py2 synapse package, which is over two years old, so I'd not expect to see upgrades from that any longer.


I think I'm happy to drop support for upgrades from our own old matrix-synapse package - as you say, it's over two years since that turned into a transitional package.

I also wanted to check that the `matrix-synapse` bug was fixed in downstream debian (since their packaging was forked from our own), but it looks like it was: https://salsa.debian.org/matrix-team/matrix-synapse/-/commit/d11ec18d6b5694e5573c152a3fddda817f6ced50#8756c63497c8dc39f7773438edf53b220c773f67_7_7

In short: fair enough, let's drop this!

File: docker/Dockerfile-dhvirtualenv
Comment (richvdh) [reply docker/Dockerfile-dhvirtualenv:56]: to be clear: it was here to help iteration when working on the first half of the Dockerfile (where it builds dh-virtualenv).

Yes it will be faster in the full-rebuild case, but not when iterating on that.

Anyway, it's all fine. Just thought I'd share the background.

File: docker/Dockerfile-dhvirtualenv
Comment (richvdh) [reply docker/Dockerfile-dhvirtualenv:27]: I'm surprised that the old incantation didn't just work. 

anyway, fair enough if you want to go this way. It just always feels wrong to me when you have to manually keep a list of packages in sync.

File: debian/matrix-synapse-py3.triggers
Comment (callahad) [reply debian/matrix-synapse-py3.triggers:7]: I would say we could generate this based on our Trove classifiers, [but...](https://github.com/matrix-org/synapse/pull/9782) 😉

File: debian/changelog
Comment (richvdh) [inline debian/changelog:4]: this needs moving to the right version, since the world has moved on in the last couple of months

**Review by richvdh — APPROVED** [review]
I think we might as well merge this rather than get hung up on the fine details. OTOH the changelog is now out of date.

lgtm once the changelog is updated.

**Comment by callahad** [comment]
(Holding this back until the 1.36 RC is branched; I'd like to couple this PR with one to actually build debs in CI)

**Comment by erikjohnston** [comment]
1.36 has been both branched and released

**Comment by erikjohnston** [comment]
@callahad is this still on your radar?

**Comment by richvdh** [comment]
I think much of this got re-done for #10429.

**Comment by richvdh** [comment]
I'm going to close this, because I think most of it is redundant now.

**Pull request closed without merging**
